### ISSUE TITLE
Regression in enumeratum support after 1.1.3

### ISSUE THREAD

**Issue by senia-psm** [softwaremill/tapir#2610]
Tapir version: 1.2.3

Scala version: 2.13.10

In version `1.1.3` one can combine these 2 endpoints:

`endpoint.in(path[MyEnumeratumEnum]("enum"))`
`endpoint.in("constant_string")`

In the latest version the result depends on the endpoint order.

Regression in `TapirCodecEnumeratum`. Introduced in this commit: https://github.com/softwaremill/tapir/commit/3810232c865eddadf2d91e34b2fcc9d5af46af8d

**How to reproduce?**

```scala
package example

import zio.*
import zio.http.{Path, Request, URL}
import zio.test.*

import enumeratum.*
import enumeratum.EnumEntry.Snakecase
import sttp.tapir.codec.enumeratum.TapirCodecEnumeratum
import sttp.tapir.server.ziohttp.ZioHttpInterpreter
import sttp.tapir.ztapir.*

object ObjectEnumSpec extends ZIOSpecDefault {
  sealed trait ST extends EnumEntry with Snakecase
  object ST extends Enum[ST] with TapirCodecEnumeratum {
    val values: IndexedSeq[ST] = findValues

    case object A extends ST
    case object B extends ST
  }

  def spec = suite("ObjectEnumSpec") {
    test("test") {
      val e1 = endpoint.in(path[ST]("st"))
      val e2 = endpoint.in("prefix")

      val r1 = e1.zServerLogic(_ => ZIO.unit)
      val r2 = e2.zServerLogic(_ => ZIO.unit)

      val r = ZioHttpInterpreter().toHttp(List(r1, r2))

      val request = Request.get(URL(Path.decode("/prefix")))

      for {
        resp <- r(request)
        body <- resp.body.asString
      } yield assertTrue(body.isEmpty)
    }
  }
}
```

Result:
```
    body = "Invalid value for: path parameter st (expected value to be one of (a, b), but got: \"prefix\")"
```

**Comment by adamw** [comment]
Thanks for the bug report.

In fact, the previous behavior was incorrect - the right result to return when the given value is not in the enumeration's allowed values is `InvalidValue`.

However, I see that changing this might also change the behavior of returning a 400 / trying the next endpoint. The relevant code in `DecodeFailureHandler` is here: https://github.com/softwaremill/tapir/blob/e1b6492af9f022cdfbc4880a6533ff4fc06b2ff2/server/core/src/main/scala/sttp/tapir/server/interceptor/decodefailure/DecodeFailureHandler.scala#L117

As a fix, you can change the `badRequestOnPathInvalidIfPathShapeMatches` from `true` to `false`, so that if a path segment fails validation, the next endpoint will be decoded (using:

```
serverOptions.decodeFailureHandler(DefaultDecodeFailureHandler(
    respond(_, badRequestOnPathErrorIfPathShapeMatches = false, badRequestOnPathInvalidIfPathShapeMatches = false),
    FailureMessages.failureMessage,
    failureResponse
  )
)
```

)

I'm afraid that changing the defaults in `DecodeFailureHandler` would be even more disruptive so I think the work-around is the best we can do currently.

**Comment by senia-psm** [comment]
@adamw could you please give me a link to some docs with explanation why `DecodeResult.Error` should allow to try the next endpoint and `DecodeResult.InvalidValue` - should not? I can't understand the difference at the moment.
If there is a reason for that - maybe the correct solution would be to add a boolean flag to `InvalidValue`?

The difference between `path[Int]` and `path[SealedTrait]` looks totally like a bug for me.

**Comment by adamw** [comment]
@senia-psm yeah, you might be right :) I don't remember why this was implemented this way to be honest - maybe I'll remind myself during the course of the day, otherwise I guess we can flip this in a minor release (1.3)

**Comment by adamw** [comment]
@senia-psm The behaviour described above was implemented as part of https://github.com/softwaremill/tapir/issues/141
See the reasoning provided there. Ordering endpoints does matter, and that was always the case. However I agree that treating `InvalidValue` and `Error` differently is surprising. So to minimise that, I'm leaning towards returning a `400` if the path shape matches, but there's any decode failure on a path segment. This would be still be configurable, of course.

**Comment by senia-psm** [comment]
@adamw changing `Error` handling might cause issues on quite common use case:
```
POST /entity/{id}
POST /entity/find
```

**Comment by adamw** [comment]
If `id` is an `Int` for example? Yeah, then you'd have to swap the ordering of these endpoints.

There is an inconsistency unfortunately and it seems to be hard to fix to suit everyone :)

**Comment by adamw** [comment]
@senia-psm After going back and forth on this, I've decided to leave the defaults as-is. I agree that it's an inconsistency. However, changing both to `false` (that is, decode error and validation failure cause a bad request) or both to `true` (always try next endpoint) would potentially silently break even more projects than we did by fixing the enumeration validation. So in a major release, I'd probably flip both to `false`. But for now, since there's a work-around, I'll keep the code unchanged.

**Comment by adamw** [comment]
I've created a PR which allows overriding the default behavior on a per-input/output basis - please take a look
